From 567eea33942dae332a4d07419faf42bf350db8c0 Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Wed, 10 Aug 2022 16:51:30 +0200
Subject: [PATCH 1/2] Round SourceBuffer removal range

Round SB removal range provided as 'double' in SB::remove() func
to 1/1000 us.

This helps to avoid cases when unintended frame
would be marked for removal becasue of very small overlap
that could come from precision lost.

It fixes HBOMax issue that uses next sample PTS as removal 'end'
and unexpectedly removes the next segment (as it's a sync frame)
---
 Source/WebCore/Modules/mediasource/SourceBuffer.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/Source/WebCore/Modules/mediasource/SourceBuffer.cpp b/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
index e546d81180ca..bbb49d3c63db 100644
--- a/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
+++ b/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
@@ -336,7 +336,10 @@ ExceptionOr<void> SourceBuffer::abort()
 
 ExceptionOr<void> SourceBuffer::remove(double start, double end)
 {
-    return remove(MediaTime::createWithDouble(start), MediaTime::createWithDouble(end));
+    // Limit timescale to 1/1000 of microsecond so samples won't accidentally overlap
+    // with removal range by precision lost (e.g. by 0.000000000000X [sec])
+    static const uint32_t timescale = 1000000000;
+    return remove(MediaTime::createWithDouble(start, timescale), MediaTime::createWithDouble(end, timescale));
 }
 
 ExceptionOr<void> SourceBuffer::remove(const MediaTime& start, const MediaTime& end)
-- 
2.25.1


From a028bb8c226bd6dc7d7340c32050c3794b5589ae Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Wed, 10 Aug 2022 16:59:33 +0200
Subject: [PATCH 2/2] Improve logging of sample data to match the rest of MSE
 logs

Use sample.toJSONString() func to log sample data as toFloat()
may print slightly different value because of limitted precision.
---
 .../platform/graphics/gstreamer/mse/PlaybackPipeline.cpp     | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/PlaybackPipeline.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/PlaybackPipeline.cpp
index 3422c45480fd..f1fc0dd1d59e 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/PlaybackPipeline.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/PlaybackPipeline.cpp
@@ -387,9 +387,8 @@ void PlaybackPipeline::enqueueSample(Ref<MediaSample>&& mediaSample)
 
     AtomString trackId = mediaSample->trackID();
 
-    GST_TRACE("enqueing sample trackId=%s PTS=%f presentationSize=%.0fx%.0f at %" GST_TIME_FORMAT " duration: %" GST_TIME_FORMAT,
-        trackId.string().utf8().data(), mediaSample->presentationTime().toFloat(),
-        mediaSample->presentationSize().width(), mediaSample->presentationSize().height(),
+    GST_TRACE("enqueing sample trackId=%s %s at %" GST_TIME_FORMAT,
+        trackId.string().utf8().data(), mediaSample->toJSONString().utf8().data(), 
         GST_TIME_ARGS(WebCore::toGstClockTime(mediaSample->presentationTime())),
         GST_TIME_ARGS(WebCore::toGstClockTime(mediaSample->duration())));
 
-- 
2.25.1

